core/iwasm/common/wasm_shared_memory.c

```diff
@@ -440,48 +427,46 @@ wasm_runtime_atomic_wait(WASMModuleInstanceCommon *module, void *address,
 
         if (0 != os_mutex_init(&wait_node->wait_lock)) {
             wasm_runtime_free(wait_node);
-            os_mutex_unlock(&wait_info->wait_list_lock);
             return -1;
         }
 
         if (0 != os_cond_init(&wait_node->wait_cond)) {
             os_mutex_destroy(&wait_node->wait_lock);
             wasm_runtime_free(wait_node);
-            os_mutex_unlock(&wait_info->wait_list_lock);
             return -1;
         }
 
         wait_node->status = S_WAITING;
-
+        os_mutex_lock(&wait_info->wait_list_lock);
         ret = bh_list_insert(wait_info->wait_list, wait_node);
+        os_mutex_unlock(&wait_info->wait_list_lock);
         bh_assert(ret == BH_LIST_SUCCESS);
         (void)ret;
     }
 
-    os_mutex_unlock(&wait_info->wait_list_lock);
-
     /* condition wait start */
     os_mutex_lock(&wait_node->wait_lock);
 
     os_cond_reltimedwait(&wait_node->wait_cond, &wait_node->wait_lock,
                          timeout < 0 ? BHT_WAIT_FOREVER
                                      : (uint64)timeout / 1000);
 
+    is_timeout = wait_node->status == S_WAITING ? true : false;
     os_mutex_unlock(&wait_node->wait_lock);
 
-    /* Check the wait node status */
+    /* Unlocked in release_wait_info() */
```

Comment:
Code may be not easy to understand if releasing the lock in release_wait_info(), and the lock/unlock order is not clear (lock1, lock2, unlock1, unlock2, but not lock1, lock2, unlock2, unlock1), how about moving the `os_mutex_unlock(&wait_info->wait_list_lock)` to be after L472:
```C
    os_mutex_lock(&wait_info->wait_list_lock);

    check_ret = is_wait_node_exists(wait_info->wait_list, wait_node);
    bh_assert(check_ret);
    ...

    os_mutex_lock(&node->shared_mem_lock);
    release_wait_info(wait_map, wait_info, address);
    os_mutex_unlock(&node->shared_mem_lock);

    os_mutex_unlock(&wait_info->wait_list_lock)
```

Reply:
The problem there is that `release_wait_info()` may call `destroy_wait_info()`
https://github.com/eloparco/wasm-micro-runtime/blob/65ce28c8085eb45d8a3032062658b35175ee2100/core/iwasm/common/wasm_shared_memory.c#L364
 so when you call `os_mutex_unlock(&wait_info->wait_list_lock)` the wait info is invalid.

I refactored the code to address that.

Reply:
OK, that looks better.